crates/sui-graphql-rpc/src/extensions/timeout.rs

```diff
@@ -2,35 +2,69 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use async_graphql::{
-    extensions::{Extension, ExtensionContext, ExtensionFactory, NextRequest},
-    Response, ServerError,
+    extensions::{Extension, ExtensionContext, ExtensionFactory, NextExecute, NextParseQuery},
+    parser::types::ExecutableDocument,
+    Response, ServerError, ServerResult,
 };
-use std::sync::Arc;
+use async_graphql_value::Variables;
 use std::time::Duration;
-use tokio::time::timeout;
+use std::{net::SocketAddr, sync::Arc};
+use tokio::{sync::Mutex, time::timeout};
+use tracing::error;
+use uuid::Uuid;
 
-use crate::config::ServiceConfig;
+use crate::{config::ServiceConfig, error::code};
 
-#[derive(Clone, Debug, Default)]
-pub(crate) struct Timeout;
+#[derive(Debug, Default)]
+pub(crate) struct Timeout {
+    pub query: Mutex<Option<String>>,
+}
 
 impl ExtensionFactory for Timeout {
     fn create(&self) -> Arc<dyn Extension> {
-        Arc::new(Timeout)
+        Arc::new(Timeout {
+            query: Mutex::new(None),
+        })
     }
 }
 
 #[async_trait::async_trait]
 impl Extension for Timeout {
-    async fn request(&self, ctx: &ExtensionContext<'_>, next: NextRequest<'_>) -> Response {
+    async fn parse_query(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        query: &str,
+        variables: &Variables,
+        next: NextParseQuery<'_>,
+    ) -> ServerResult<ExecutableDocument> {
+        let document = next.run(ctx, query, variables).await?;
+        *self.query.lock().await = Some(ctx.stringify_execute_doc(&document, variables));
+        Ok(document)
+    }
+
+    async fn execute(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        operation_name: Option<&str>,
+        next: NextExecute<'_>,
+    ) -> Response {
         let cfg = ctx
             .data::<ServiceConfig>()
             .expect("No service config provided in schema data");
+        let query = self.query.lock().await.take().unwrap_or("".to_string());
```

Comment:
The use of `take` here is a bit strange here -- why are we taking it instead of borrowing it, and why are we doing it here vs once we know that an error has happened?

Reply:
Regarding the 2nd part of the question, putting this into the `unwrap_or_else` closure will not work due to the need for calling `await`.

Reply:
Switching to std::sync::Mutex fixed this.